pkg/monitor/monitor.go

```diff
@@ -280,32 +298,76 @@ func (m *Monitor) cleanupDriverNodePods(node *volume.NodeInfo) {
+	csiPodsSelectedForDeletion := make([]v1.Pod, 0)
 	for _, pod := range pods.Items {
-		var msg string
 		csiPodPrefix, err := m.Driver.GetCSIPodPrefix()
 		if err == nil && strings.HasPrefix(pod.Name, csiPodPrefix) {
-			msg = fmt.Sprintf("Deleting csi pod from Node %v due to volume driver status: %v (%v)", pod.Spec.NodeName, node.Status, node.RawStatus)
+			csiPodsSelectedForDeletion = append(csiPodsSelectedForDeletion, pod)
 		} else {
 			if len(pod.Spec.Volumes) == 0 {
 				// Pod does not have any volumes. So, no need to check further
 				continue
 			}
-			msg = fmt.Sprintf("Deleting Pod from Node %v due to volume driver status: %v (%v)", pod.Spec.NodeName, node.Status, node.RawStatus)
 			owns, err := m.doesDriverOwnPodVolumes(&pod)
 			if err != nil || !owns {
 				continue
 			}
+			podsSelectedForDeletion = append(podsSelectedForDeletion, pod)
+		}
+	}
+
+	// Delete CSI pods first
+	if len(csiPodsSelectedForDeletion) > 0 {
+		m.batchDeleteOfPodsFromOfflineNodes(csiPodsSelectedForDeletion, node, true)
+	}
+
+	// Do batch deletes of rest of the pods
+	if len(podsSelectedForDeletion) > 0 {
+		m.batchDeleteOfPodsFromOfflineNodes(podsSelectedForDeletion, node, false)
+	}
+}
+
+// batchDeleteOfPodsFromOfflineNodes deletes the pods in batches of batchSize
+// and increments the HealthCounter by the number of pods deleted
+func (m *Monitor) batchDeleteOfPodsFromOfflineNodes(
+	pods []v1.Pod,
+	node *volume.NodeInfo,
+	csiPods bool) {
+
+	// Do a fresh node status check before deleting pods
+	n, err := m.Driver.InspectNode(node.StorageID)
+	if err != nil {
+		log.Errorf("Error inspecting node %v in batch deletion of pods: %v", node.StorageID, err)
+		return
+	}
+	if !m.isNodeOffline(n) {
+		log.Infof("Volume driver on node %v (%v) is still offline (%v) in batch deletion of pods", node.Hostname, node.StorageID, n.RawStatus)
+		return
+	}
+	log.Infof("Going to delete %d number of pods from node %v in batch", len(pods), node.Hostname)
+	for i := 0; i < len(pods); i += podDeleteBatchSize {
+		end := i + podDeleteBatchSize
+		if end > len(pods) {
+			end = len(pods)
 		}
-		if m.isSameNode(pod.Spec.NodeName, node) {
+		counterIncrement := end - i
+		for _, pod := range pods[i:end] {
+			var msg string
+			if csiPods {
+				msg = fmt.Sprintf("Deleting CSI Pod from Node %v due to volume driver status: %v (%v)", pod.Spec.NodeName, node.Status, node.RawStatus)
+			} else {
+				msg = fmt.Sprintf("Deleting Pod from Node %v due to volume driver status: %v (%v)", pod.Spec.NodeName, node.Status, node.RawStatus)
+			}
 			storklog.PodLog(&pod).Infof(msg)
 			m.Recorder.Event(&pod, v1.EventTypeWarning, storageDriverOfflineReason, msg)
-			err := core.Instance().DeletePods([]v1.Pod{pod}, true)
-			if err != nil {
-				storklog.PodLog(&pod).Errorf("Error deleting pod: %v", err)
-				continue
-			}
-			HealthCounter.Inc()
 		}
+
+		err := core.Instance().DeletePods(pods[i:end], true)
+		if err != nil {
+			log.Errorf("Error deleting pods: %v", err)
```

Comment:
It is a known problem that PX can come online and we may continue to delete the Pods. Does it not make sense to check the status of PX after each batch?

Reply:
Good suggestion, now moved the Inspectnode into each batch, also added a default 30s interval between batches of pod deletes.